File: projects/Bouncy Ball/scripts/Missile.lua
BlankE.addEntity("Missile")

function Missile:init()
	-- add missile image
	self:addSprite{name="main",image="missile",align="center"}
	-- make the missile non-homing when it spawns
	self.homing = false
	-- have the missile fade in while it's "charging up"
	self.sprite_alpha = 0
	-- at the end of the tween, set the missile to homing
	Tween(self, {sprite_alpha=1}, 1, "linear", function()
		-- at the end of the tween, set the missile to homing
		self.homing = true
		Timer.after(10, function()
			self.homing = false
		end)
	end):play()
	
	-- give it a hitbox
	self:addShape("main","rectangle")
end

function Missile:explode()
	-- break the missile image into pieces
	self.img_missile_pcs = self:getImage():chop(5,5)
	-- throw them in the opposite direction 
	local opp_direction = self.sprite_angle + 180
	self.img_missile_pcs:forEach(function(piece)
		local direction = randRange(opp_direction - 45, opp_direction + 45)
		piece.hspeed = direction_x(direction, 20)
		piece.vspeed = direction_y(direction, 20)
	end)
	-- draw the pieces
	if self.img_missile_pcs then
		self.img_missile_pcs:call('draw')
	end
	-- remove the hitbox so that it seems like it's actually exploded and gone
	self:removeShape("main")
	-- after some time, destroy the object
	Timer.after(3, function()
		self:destroy()
	end)
end

function Missile:update(dt)
	-- rotate the image to match the direction it's moving in
	self.sprite_angle = self.direction + 90
	
	-- if there is a paddle, move towards it
	local paddle = Paddle.instances[1]
	if paddle and self.homing then
		self:moveTowards(paddle, 100)
	end
	
	-- call our custom explode() method during a collision
	self.onCollision["main"] = function(other)
		if other.parent.classname == "Paddle" and self.homing then
			self:explode()
			other.parent:explode()
		end
	end	
end

function Missile:draw()
	-- draw the pieces
	-- hide the old image if it's exploded
	if self.img_missile_pcs then
		self.img_missile_pcs:call('draw')
	else
		self:drawSprite()
	end
end

File: old_projects/tut1-bouncyball/scripts/Missile.js
class Missile extends Entity {
    init () {
		this.addSprite("missile")
		this.sprite_align = "center"
		// after 10 seconds, stop homing in on the Paddle
		this.homing = false;
		Timer.after(10, ()=>{
			this.homing = false;
		})
		// wait 3 seconds before moving
		this.sprite_alpha = 0
		let twn = new TWEEN.Tween(this)
			.to({ sprite_alpha:1 },3000)
			.easing(TWEEN.Easing.Quadratic.Out)
			.onComplete(()=>{
				this.homing = true;
				this.addShape('main','circle')
			})
			.start()
		
    }
    update (dt) {
		this.sprite_angle = this.direction + 90;
		let paddle = Paddle.instances[0];
		
		if (!paddle) {
			this.homing = false
		}
		if (this.homing) {
			this.moveTowards(paddle.x, paddle.y, 1)
		}
    }
}

TestScene({
	onStart () {
		let pad = new Paddle()
		pad.x = 150
		pad.y = 150
		let m = new Missile()
		m.x = 100
		m.y = 200
	}
})